File: funnyvote/app/src/main/java/com/heaton/funnyvote/ui/createvote/CreateVoteTabOptionFragment.java
package com.heaton.funnyvote.ui.createvote;

import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.heaton.funnyvote.R;
import com.heaton.funnyvote.database.Option;
import com.theartofdev.edmodo.cropper.CropImage;

import java.util.List;

/**
 * Created by heaton on 2016/9/1.
 */

public class CreateVoteTabOptionFragment extends Fragment implements CreateVoteContract.OptionFragmentView {
    private RecyclerView ryOptions;
    private View rootView;
    private OptionCreateItemAdapter optionItemAdapter;
    private OptionItemListener itemListener;
    private CreateVoteContract.Presenter presenter;
    ImageView imgMain;
    ImageView imgPick;

    public CreateVoteTabOptionFragment() {
    }

    public static CreateVoteTabOptionFragment newTabFragment() {
        return new CreateVoteTabOptionFragment();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_create_vote_tab_options, container, false);
        ryOptions = (RecyclerView) rootView.findViewById(R.id.ryOptions);
        imgPick = (ImageView) rootView.findViewById(R.id.imgPick);
        imgMain = (ImageView) rootView.findViewById(R.id.imgMain);

        itemListener = new OptionItemListener() {
            @Override
            public void onOptionTextChange(long optionId, String newOptionText) {
                presenter.reviseOption(optionId, newOptionText);
            }

            @Override
            public void onOptionAddNew() {
                presenter.addNewOption();
            }

            @Override
            public void onOptionRemove(long optionId) {
                presenter.removeOption(optionId);
            }
        };

        View.OnClickListener pickImageListener = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                CropImage.startPickImageActivity(getActivity());
            }
        };
        imgMain.setOnClickListener(pickImageListener);
        imgPick.setOnClickListener(pickImageListener);
        return rootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        presenter.setOptionFragmentView(this);
    }

    @Override
    public void setUpOptionAdapter(List<Option> optionList) {
        optionItemAdapter = new OptionCreateItemAdapter(optionList, itemListener);
        ryOptions.setAdapter(optionItemAdapter);
    }

    @Override
    public void setVoteImage(Uri imageUri) {
        imgMain.setVisibility(View.VISIBLE);
        imgPick.setVisibility(View.GONE);
        Glide.with(this)
                .load(imageUri)
                .into(imgMain);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
    }

    @Override
    public void setPresenter(CreateVoteContract.Presenter presenter) {
        this.presenter = presenter;
    }

    @Override
    public void refreshOptions() {
        optionItemAdapter.notifyDataSetChanged();
    }

    public interface OptionItemListener {
        void onOptionTextChange(long optionId, String newOptionText);

        void onOptionAddNew();

        void onOptionRemove(long optionId);
    }
}

File: funnyvote/app/src/main/java/com/heaton/funnyvote/ui/createvote/CreateVoteTabOptionFragment.kt
package com.heaton.funnyvote.ui.createvote

import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.bumptech.glide.Glide
import com.heaton.funnyvote.database.Option
import com.heaton.funnyvote.databinding.FragmentCreateVoteTabOptionsBinding
import com.theartofdev.edmodo.cropper.CropImage

class CreateVoteTabOptionFragment : Fragment(), CreateVoteContract.OptionFragmentView {
    private var _binding: FragmentCreateVoteTabOptionsBinding? = null
    private val binding get() = _binding!!

    private var optionItemAdapter: OptionCreateItemAdapter? = null
    private var itemListener: OptionItemListener? = null
    private lateinit var presenter: CreateVoteContract.Presenter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentCreateVoteTabOptionsBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        itemListener = object : OptionItemListener {
            override fun onOptionTextChange(optionId: Long, newOptionText: String) {
                presenter.reviseOption(optionId, newOptionText)
            }

            override fun onOptionAddNew() {
                presenter.addNewOption()
            }

            override fun onOptionRemove(optionId: Long) {
                presenter.removeOption(optionId)
            }
        }

        val pickImageListener = View.OnClickListener { CropImage.startPickImageActivity(requireActivity()) }
        binding.imgMain.setOnClickListener(pickImageListener)
        binding.imgPick.setOnClickListener(pickImageListener)
        presenter.setOptionFragmentView(this)
    }

    override fun setUpOptionAdapter(optionList: List<Option>) {
        _binding?.let { b ->
            optionItemAdapter = OptionCreateItemAdapter(optionList, itemListener!!)
            b.ryOptions.adapter = optionItemAdapter
        }
    }

    override fun setVoteImage(imageUri: Uri) {
        _binding?.let { b ->
            b.imgMain.visibility = View.VISIBLE
            b.imgPick.visibility = View.GONE
            Glide.with(this)
                .load(imageUri)
                .into(b.imgMain)
        }
    }

    override fun setPresenter(presenter: CreateVoteContract.Presenter) {
        this.presenter = presenter
    }

    override fun refreshOptions() {
        optionItemAdapter?.notifyDataSetChanged()
    }

    interface OptionItemListener {
        fun onOptionTextChange(optionId: Long, newOptionText: String)
        fun onOptionAddNew()
        fun onOptionRemove(optionId: Long)
    }

    companion object {
        fun newTabFragment(): CreateVoteTabOptionFragment {
            return CreateVoteTabOptionFragment()
        }
    }
}
